src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2168,162 +2168,151 @@ cdef class Polynomial(CommutativePolynomial):
-                    if hd != 0 and hd != self.degree():
-                        if 2*hd <= self.degree():
-                            return h.any_root(ring, -degree, True)
+
+                    # If we found a degree-one root, return it
+                    if hd.is_one():
+                        return _linear_root(h, ring=ring) 
+
+                    # Else check if we have a non-trivial factor and keep going
+                    if not hd.is_zero() and hd != f.degree():
+                        if 2*hd <= f.degree():
+                            return _cantor_zassenhaus_split_root(h, degree, ring=ring)
                         else:
-                            return (self//h).any_root(ring, -degree, True)
-        else:
-            rs = self.roots(ring=ring, multiplicities=False)
-            if rs:
-                return rs[0]
-            raise ValueError("no roots (non-field) %s" % self)
+                            return _cantor_zassenhaus_split_root(f//h, degree, ring=ring)
+                # If you are reaching this error, chances are there's a bug in the code.
+                raise ValueError(f"no splitting of degree {degree} found for {f}")
+
+            # Now handle even charactertistic
+            for _ in range(1000):
+                # Sample a uniformly random element of R
+                T = R.random_element(2*degree + 1)
+                if T.is_zero(): continue
+                T = T.monic()
+
+                # Compute the trace of T with field of order 2^k
+                # sum T^(2^i) for i in range (degree * k)
+                # We use repeated squaring to avoid redundent multiplications
+                C, TT = T, T
+                for _ in range(degree * f.base_ring().degree() - 1):
+                    TT = pow(TT, 2, f)
+                    C += TT
+
+                # Compute the gcd to find a factor
+                h = f.gcd(C)
+                hd = h.degree()
+
+                # If we found a degree-one root, return it
+                if hd.is_one():
+                    return _linear_root(h, ring=ring)
+
+                # else check if we have a factor
+                elif not hd.is_zero() and hd != f.degree():
+                    if 2*hd <= f.degree():
+                        return _cantor_zassenhaus_split_root(h, degree, ring)
+                    else:
+                        return _cantor_zassenhaus_split_root(f // h, degree, ring)
+
+            # If you are reaching this error, chances are there's a bug in the code.
+            raise ValueError(f"no splitting of degree {degree} found for {f}")
+        # ======================== #
+        #   End helper functions   #
+        # ======================== #
+
+        # Initial checks for bad input
+        if self.degree() < 0:
+            if ring is None:
+                return self.base_ring()(0)
+            return ring(0)
+        elif self.degree() == 0:
+            raise ValueError(f"no root can be computed for {self}")
```

Comment:
We may want to point out to the user *why* we can't compute a root — there aren't any!